src/circuit.rs

```diff
@@ -1087,6 +1156,46 @@ mod tests {
         assert_eq!(proof.0.len(), expected_proof_size);
     }
 
+    #[test]
+    fn test_not_equal_asset_ids() {
+        let mut rng = OsRng;
+
+        let (mut circuits, instances): (Vec<_>, Vec<_>) = iter::once(())
+            .map(|()| generate_circuit_instance(&mut rng))
+            .unzip();
+
+        // Modify asset_new to not be equal to asset_old
+        let random_asset_id = {
+            let sk = SpendingKey::random(&mut rng);
+            let isk = IssuanceAuthorizingKey::from(&sk);
+            let ik = IssuanceValidatingKey::from(&isk);
+            let asset_descr = "zsa_asset";
+            AssetId::derive(&ik, asset_descr)
+        };
+        circuits[0].asset_new = Value::known(random_asset_id);
+
+        let vk = VerifyingKey::build();
+
+        for (circuit, instance) in circuits.iter().zip(instances.iter()) {
+            assert!(MockProver::run(
+                K,
+                circuit,
+                instance
+                    .to_halo2_instance()
+                    .iter()
+                    .map(|p| p.to_vec())
+                    .collect()
+            )
+            .unwrap()
+            .verify()
+            .is_err());
```

Comment:
Should explain why we check twice. Here and again in line 1196.

Reply:
Also, a specific expected error is better. (here and line 1196)

Reply:
I prefer to keep the two checks (with mock prover and with plonk prover) in order to have the same structure than circuit::tests::round_trip test.
Only the verification with the PLONK prover is necessary

Reply:
With plonk prover, we cannot check a specific expected error